feat: implement timestamp callback plugin to show simple timestamp for each header

kurokobo opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

This PR introduce new callback plugin to show simple timestamp for each header line.

This plugin reduces the number of `*`'s in the header by the number of characters in the timestamp, and display the timestamp on the far right.

![image](https://github.com/ansible-collections/community.general/assets/2920259/f312ecb3-3702-43a9-8f01-4fc2b4e5be29)

On the output screen of an AWX job, a timestamp appears in each header by default. This PR implements this timestamp idea as a callback plugin.

![image](https://github.com/ansible-collections/community.general/assets/2920259/cb60b996-7e50-4519-9af5-cedcf2b32dff)

There is a plugin that can be used for a similar purpose, [`ansible.posix.profile_tasks`](https://docs.ansible.com/ansible/latest/collections/ansible/posix/profile_tasks_callback.html), but it adds a new line for displaying the timestamps, which complicates the output.
The plugin in this PR is as simple as possible and provides only minimal information, so it is not at all obtrusive to keep it enabled at all times.

Available options:

- `timezone`: Users can specify timezone for the timestamp. This is helpful for containerized ansible environment. **Ignored on Python < 3.9**
- `format_string`: The format string for timestamp

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

<!--- Please do not forget to include a changelog fragment:
      https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-changelog-fragments
      No need to include one for docs-only or test-only PR, and for new plugin/module PRs.
      Read about more details in CONTRIBUTING.md.
      -->

##### ISSUE TYPE
<!--- Pick one or more below and delete the rest.
      'Test Pull Request' is for PRs that add/extend tests without code changes. -->
- New Module/Plugin Pull Request

##### COMPONENT NAME
<!--- Write the SHORT NAME of the module, plugin, task or feature below. -->

timestamp

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

Tested with:

```bash
ansible-test sanity --docker -v plugins/callback/timestamp.py
ansible-test integration --docker -v callback_timestamp
```

This is the first time for me to send PR for this collection. I have a few concerns and would love to hear any input. Thanks!

- Is this plugin appropriate for community.general in the first place?
- The timezone is set using [ZoneInfo](https://docs.python.org/3/library/zoneinfo.html), which was added in Python 3.9. This means it will not work in Python 3.8. This collection [requires Ansible 2.13 or higher](https://github.com/ansible-collections/community.general/blob/main/meta/runtime.yml#L6), and [Ansible 2.13 supports Python 3.8](https://docs.ansible.com/ansible/devel/reference_appendices/release_and_maintenance.html#support-life). However, since it is already EOL, this plugin has decided to ignore the user-provided timezone setting for Python 3.8 and below, simplifying the implementation. Is this a good practice?
- About the test. I tried to write an integration test, but I cannot use the `callback` role like other callback plugins because the timestamp changes each time it is executed. For this reason I am only testing that the timestamp format is given a fixed string and that at least the timestamp is displayed in the proper position without the header line being folded back (this is the core feature of this plugin that has to be tested at least). Should I implement the test more rigorously?

If this should not continue, feel free to close it. Thanks in advance! 😃

kurokobo commented:
Fixed issues for license and type annotation that caused CI to fail. Sorry for my trivial mistakes.
CI still fails but appears to be unrelated to my PR. Thanks😃

felixfontein commented on plugins/callback/timestamp.py:
```diff
@@ -0,0 +1,111 @@
+  name: timestamp
+  type: stdout
+  short_description: Adds simple timestamp for each header
+  version_added: 9.0.0  # for collections, use the collection version, not the Ansible version
```
```suggestion
  version_added: 9.0.0
```

felixfontein commented on plugins/callback/timestamp.py:
```diff
@@ -0,0 +1,111 @@
+  short_description: Adds simple timestamp for each header
+  version_added: 9.0.0  # for collections, use the collection version, not the Ansible version
+  description:
+    - This callback adds simple timestamp for each header
```
```suggestion
    - This callback adds simple timestamp for each header.
```

felixfontein commented on plugins/callback/timestamp.py:
```diff
@@ -0,0 +1,111 @@
+  author: kurokobo (@kurokobo)
+  options:
+    timezone:
+      description: timezone to use for the timestamp in IANA time zone format (for example C(America/New_York), C(Asia/Tokyo)). Ignored on Python < 3.9
```
```suggestion
      description:
        - Timezone to use for the timestamp in IANA time zone format.
        - For example C(America/New_York), C(Asia/Tokyo)). Ignored on Python < 3.9.
```

felixfontein commented on plugins/callback/timestamp.py:
```diff
@@ -0,0 +1,111 @@
+          key: timezone
+      env:
+        - name: ANSIBLE_CALLBACK_TIMESTAMP_TIMEZONE
+      default: ""
```
```suggestion
      type: string
```
The empty string is NOT a valid time zone, so it should not be used as a default.

kurokobo replied:
TIL that if `default` is not defined here, the actual default is null

felixfontein commented on plugins/callback/timestamp.py:
```diff
@@ -0,0 +1,111 @@
+          key: format_string
+      env:
+        - name: ANSIBLE_CALLBACK_TIMESTAMP_FORMAT_STRING
+      default: "%H:%M:%S"
```
```suggestion
      default: "%H:%M:%S"
      type: string
```

felixfontein commented on plugins/callback/timestamp.py:
```diff
@@ -0,0 +1,111 @@
+        - name: ANSIBLE_CALLBACK_TIMESTAMP_TIMEZONE
+      default: ""
+    format_string:
+      description: format of the string shown to user
```
```suggestion
      description:
        - Format of the string shown to user.
```
This also needs an explanation (or a link to an explanation) of which formatting sequences can be used.

kurokobo replied:
Added the description for 1989 C standards and also added link to the python docs

felixfontein commented on plugins/callback/timestamp.py:
```diff
@@ -0,0 +1,111 @@
+    """
+    Prints a header-looking line with cowsay or stars with length depending on terminal width (3 minimum) with trailing timestamp
+
+    Based on the banner method of Display class from ansible.utils.display
```
In that case the copyright statement for this plugin should mention the copyright holders of that file (and potentially others that you copied code from).

kurokobo replied:
The `banner` method is from here: https://github.com/ansible/ansible/blob/4403519afe89138042108e237aef317fd5f09c33/lib/ansible/utils/display.py#L511

So I added the copyright from above code. The license of above code was the same GPLv3+.

felixfontein reviewed (commented):
Thanks for your contribution!

I would probably add a `seealso` reference to the ansible.posix.profile_tasks callback (https://github.com/ansible-collections/ansible.posix/blob/main/plugins/callback/profile_tasks.py), which (among other information) also adds timestamps to tasks.

kurokobo commented:
@felixfontein 
Hi, thanks for your extremely rapid review! Corrections have been completed for all your comments. Please let me know if you have any other concerns.
Thanks! 😃

kurokobo commented:
Sorry for massive updates, I've corrected invalid links (`L()` v.s. `U()`) and added `seealso`. Thanks!

kurokobo commented:
@felixfontein 
Thanks for working on this PR.
The EOL CI seems to have failed, but I believe this is intended since that adding any plugins to `seealso` is not yet supported in 2.13 and 2.14 of ansible-doc.

Any opinions on whether this failure should be ignored, or fixed in some way? Thanks!

felixfontein commented on plugins/callback/timestamp.py:
```diff
@@ -0,0 +1,127 @@
+    - plugin: ansible.posix.profile_tasks
+      plugin_type: callback
+      description: >
+        You can use ansible.posix.profile_tasks callback plugin to time individual tasks and overall execution time
```
```suggestion
        You can use P(ansible.posix.profile_tasks#callback) callback plugin to time individual tasks and overall execution time
```

felixfontein commented:
> The EOL CI seems to have failed, but I believe this is intended since that adding any plugins to `seealso` is not yet supported in 2.13 and 2.14 of ansible-doc.
> 
> Any opinions on whether this failure should be ignored, or fixed in some way? Thanks!

Please add ignore entries to `tests/sanity/ignore-2.13.txt` and `tests/sanity/ignore-2.14.txt`:
```
plugins/callback/timestamp.py validate-modules:invalid-documentation
```
(Please make sure to insert them alphabetically, i.e. before the first lookup.)

kurokobo commented:
@felixfontein 
Thanks for reviewiing this again, now I learned what `ignore-*` files are for.
Updated, and rebased (I believe rebasing is required to merge). Thanks!

felixfontein commented:
> Updated, and rebased (I believe rebasing is required to merge). Thanks!

Rebasing is not needed. It's only needed when you have conflicts, or you need code that's already in `main` but not yet in your PR's branch.

kurokobo commented:
Aww, I was confusing the rules with other repositories, sorry for annoying you 😞

felixfontein commented:
No worries! Rebasing doesn't hurt, and you definitely didn't annoy me :)

felixfontein reviewed (approved):
Looks good to me! If nobody objects, I'll merge this in ~a week.

russoz commented:
LGTM

felixfontein commented:
@kurokobo thanks for contributing this!
@russoz thanks for reviewing!
